### PR TITLE
Add CMakeLists.txt

### PR DESCRIPTION
**Pull request by AlexisMotet**
Hi,
Thanks for leaving your project open. I need an allocator for a research project in C++, I would be interested in contributing, let me know if you are ok. I'm using Visual Studio 2019 / clang / gcc.

### CODE REVIEW

**Comment by cozis** [comment]
Hey @AlexisMotet. I am open to contributions, but what kind of change did you have in mind? I'll look at this PR a bit later when I'm free

**Comment by AlexisMotet** [comment]
It would be mostly to code it in C++ 20 actually, I don't know if you're interested. 

But I think I'll try coding from scratch to get a good understanding of the algorithm. Thanks for the explanations in the code.

**Comment by cozis** [comment]
Getting this code to work with a C++ compiler would be nice! Introducing C++ specific features may be too much though. If you decide to write your own allocator let me know if the comments help or if there are any mistakes.

If you're still interested in contributing here are a few notes:

The cmake version is a bit high. You require 2.23 but the package manager on my WSL install only goes up to 2.22. I did change the version in CMakeLists.txt and it appeared to work fine, so maybe you can lower it?

The comment is slightly wrong here. Since MIN_BLOCK_LOG2 must be greater than 3, the minimum size is 16 and not 8
```c
_Static_assert(BUDDY_ALLOC_MIN_BLOCK_LOG2 > 3, "the min block must be at least 8 bytes long");
```

**Comment by AlexisMotet** [comment]
Yes, I can lower the CMake version, but this pull request was just to create a chat you can ignore it. 

I've pushed a “buddy.hpp” https://github.com/AlexisMotet/buddy/tree/main to show you the interface I'd need in C++, let me know what you think.

**Comment by cozis** [comment]
I realize the buddy_startup interface could be simplified quite a bit. Basically, for each block of size `BUDDY_MAX_BLOCK_SIZE` of the memory pool, you need to provide a `struct page_info`. Here is an example:

```c
int main()
{
    #define N 10

    char mem[BUDDY_MAX_BLOCK_SIZE * N];

    struct page_info pinfo[N];
    struct buddy_alloc alloc = buddy_startup(mem, sizeof(mem), pinfo, N);

    // ...

    buddy_cleanup(&alloc);
    return 0;
}
```
I realize now that the better API would handle everything internally using the memory provided by the user

File: buddy.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+#include "buddy.h"
+
+class BuddyAllocator {
+    private:
+        struct buddy_alloc allocator;
+        struct page_info info; // I didn't understand the point of the pages though
+    public:
+        BuddyAllocator(std::byte* memory, std::size_t size) {
+            buddy_startup(reinterpret_cast<char*>(memory), size, &info, 1);
```
Comment (cozis) [inline buddy.hpp:14]: I think you forgot to assign to `allocator` here?

File: buddy.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        ~BuddyAllocator() {
+            buddy_cleanup(&allocator);
+        }
+
+        std::byte* allocate(std::size_t size) {
+            auto* ptr = reinterpret_cast<std::byte*>(buddy_malloc(&allocator, size));
+            if (ptr == nullptr) {
+                throw std::runtime_error("todo");
```
Comment (cozis) [inline buddy.hpp:24]: This is something that may require some thought. Throwing needs heap memory, so throwing here would add a dependency of the library to the global allocator. This is probably not good

**Comment by AlexisMotet** [comment]
For pages: what I don't understand is why you need several “trees” of blocks/pages? Can't you just build a large tree from the given memory size? is it when the memory supplied is really too big?

For the exception: Ok then it can just return nullptr.

Also, in C++ 20 there are several functions that could be used for dealing with bits: 
https://en.cppreference.com/w/cpp/header/bit

round_pow2 <=> std::bit_ceil for example
first_set <=> std::bit_width ? I'm not sure

File: buddy.hpp (hunk 1 above)
Comment (AlexisMotet) [reply buddy.hpp:14]: yes

**Comment by cozis** [comment]
Each page has its own tree (and by page I actually mean an aligned block of size BUDDY_MAX_BLOCK_SIZE). If you only provide it one bit tree, the allocator will only consider the first aligned page of the pool. I think I will change the API.

It's important to me that the actual buddy.c and buddy.h can be compiled with a C compiler. I don't mind adding a new header file for C++ or doing compatibility fixes (eg avoid implicit void conversions), but adding C++ features to the main code I would avoid.

You could always make the code C++20 in your own fork. I'd be curious to see how it turns out.

**Comment by cozis** [comment]
I simplified the API quite a bit. You may want to rebase now

**Comment by AlexisMotet** [comment]
Hi, finally I've started from scratch because I can afford to use malloc/heap allocation in the algorithm even if it looks weird. I just need an allocator on a segment, I have no other constraints. I don't think this is a very common use case.

https://github.com/AlexisMotet/BuddyAllocator/blob/main/buddy.hpp

I did it in C++, I think I may compare the performance with your implementation tonight, I'll let you know.

**Comment by cozis** [comment]
ok :)

**Comment by AlexisMotet** [comment]
I've added this file to compare performance. 

https://github.com/AlexisMotet/BuddyAllocator/blob/main/perf.cpp

I think your code has a problem for now because it's below C++, but it can't be anything serious. But I think that in the end, C++ should be barely slower than C. And the buddy allocator C++ is just below malloc, as it doesn't have to deal with concurrency I guess ?

image
(Release + O2 flags, Windows, Intel(R) Xeon(R) W-2123 CPU @ 3.60GHz, 32Go)

I've tried to compare with another implementation https://github.com/red-rocket-computing/buddy-alloc. 
... But I've got bugs for now. I was trying on Windows but I'll test with gcc tonight (I don't have WSL). They embed the allocator so it seems more prone to bugs.

For the next I'll take a look at this paper I think, to see if it can be improved:
Fast Allocation and Deallocation with an Improved Buddy System
https://cs.au.dk/~gerth/papers/actainformatica05.pdf

**Comment by cozis** [comment]
Hey, thanks for the feedback! I'll try timing the code to see what's going on.

The paper looks interesting. If you're going to work on this feel free to ping me in pvt. I'm always interested in talking about this kind of stuff

**Pull request closed without merging**